### DISCUSSION TITLE
Error message when creating Developer Key On Canvas

### DISCUSSION THREAD

**Discussion by yongra** [tl-its-umich-edu/my-learning-analytics#1311] [Q&A]
I follow the direction on readme carefully, but I think I am missing some steps. When I create a new LTI developer key on Canvas, I get "Request failed with status code 422." When I use the JSON URL to a web browser, I get the following error. 

{"lti_error": "Invalid LTI configuration: \"[Errno 2] No such file or directory: '/secrets/private.key'\""}

I must be missing some configuration on the env.json file for the LTI coniguration section. I am not sure what parts I need to make changes to work with out test server for now. Not ready to deploy this to production yet.

"/*  LTI 1.3 configuration":"*/",
    "/*  The first key of LTI_CONFIG is the Canvas URL (production, beta, or test)":"*/",
    "LTI_CONFIG": {
        "https://canvas.instructure.com": [
            {
                "/*The default set of variables for LTI validation":"*/",
                "default": true,
                "/*LTI Dev Key from Canvas":"*/",
                "client_id": "17700000000000111",
                "/*Allowed hosts for the following 3 URLs: canvas.instructure.com, canvas.beta.instructure.com, canvas.test.instructure.com":"*/",
                "auth_login_url": "https://canvas.instructure.com/api/lti/authorize_redirect",
                "auth_token_url": "https://canvas.instructure.com/login/oauth2/token",
                "key_set_url": "https://canvas.instructure.com/api/lti/security/jwks",
                "key_set":null,
                "/*Tools private key for LTI validation":"*/",
                "private_key_file": "/secrets/private.key",
                "/*Tools public key for LTI validation":"*/",
                "public_key_file": "/secrets/public.key",
                "/*Tools installation id in platform":"*/",
                "deployment_ids": ["27297:7db438071375c02373713c12c73869ff2f470b68"]
            }
        ]
    },

Can anyone tell me what parts I need to update?

Thank you.

**Comment by yongra** [comment]
I got through some of the blocks. Now I am getting this error.

lti_error | "LTI Login failed due to missing \"target_link_uri\" param"


I am not sure what parameter is needed for this. It seems to look OK on the developer key settings.

**Reply by pushyamig** [reply]
I don't know how you did the set up, the error is coming from this [line](https://github.com/tl-its-umich-edu/my-learning-analytics/blob/master/dashboard/lti_new.py#L270)  

Make sure the LTI launch URL is configured correctly. Something similar to this
![Screen Shot 2021-08-26 at 12 16 52 PM](https://user-images.githubusercontent.com/8579775/130999431-78a468b3-9ca8-4fef-a53a-40aae1c47169.png)

**Reply by yongra** [reply]
I checked and they are all correct with our server's FQDN.

My guess is the env.json file's LTI configuration. 

For the following part, do I need to replace it with our Canvas URL? My guess is leave it as it is.

"/Allowed hosts for the following 3 URLs: canvas.instructure.com, canvas.beta.instructure.com, canvas.test.instructure.com":"/",
"auth_login_url": "https://canvas.instructure.com/api/lti/authorize_redirect",
"auth_token_url": "https://canvas.instructure.com/login/oauth2/token",
"key_set_url": "https://canvas.instructure.com/api/lti/security/jwks",

Also, are there any other settings I need to check/update?

Ra

**Reply by pushyamig** [reply]
The above URL should reflect your canvas environment where you are trying to Run MyLA. For eg., if you are running the Myla from canvas test then it should be `canvas.test.instructure.com` Make sure that is true for all urls's in LTI_CONFIG object

Also set `"CSRF_TRUSTED_ORIGINS": ["instructure.com"],` if your institution canvas URL ends like that
set `"CSRF_COOKIE_SECURE": true,`. --although the error you are seeing not because of this.

**Reply by yongra** [reply]
I double-check all the settings, but still no dice. I must be missing something. Do you have any documentation on what needs to be edited on the config file? Is env.json is the only file that needs to be configured? Google Chrome just tells me refuse to connect. Firefox is the one that gave me little more information. 

Do I need to set up the SAML as well?

**Comment by pushyamig** [comment]
> Do I need to set up the SAML as well?

No.
Below are some configuration for LTI to work
   ```
 "STUDENT_DASHBOARD_LTI": true,
    "CSRF_COOKIE_SECURE": true,
    "CSRF_TRUSTED_ORIGINS": ["instructure.com"],
    "SESSION_COOKIE_SAMESITE": "None",
    "CSRF_COOKIE_SAMESITE": "None",
    "ENABLE_BACKEND_LOGIN": false,
    "STUDENT_DASHBOARD_SAML": false,
"LTI_CONFIG": {
        "https://canvas.test.instructure.com": [
            {
                "default": true,
                "client_id": "17700000000000001",
                "auth_login_url": "https://canvas.test.instructure.com/api/lti/authorize_redirect",
                "auth_token_url": "https://canvas.test.instructure.com/login/oauth2/token",
                "key_set_url": "https://canvas.test.instructure.com/api/lti/security/jwks",
                "key_set":null,
                "private_key_file": "/secrets/private.key",
                "public_key_file": "/secrets/public.key",
                "deployment_ids": ["164:3869ff2f470b68"]
            }
        ]
    },
```

I will try to run myself and see if I get any errors.

**Reply by yongra** [reply]
Thanks for the sample. I compared it with my configuration and it matches. But, still have the issue.

**Reply by pushyamig** [reply]
I ran locally and i don't see any issues launching the MyLA locally. I have sample env.json in this document
https://docs.google.com/document/d/1WBdHR5Iz2u2Mcucp4CrgThrhBnuq7XR3Q1UgE6j6BX8/edit?usp=sharing

Note: that all the DB connection credentials needs to be replaced with your and LTI_CONFIG as well.

**Reply by yongra** [reply]
The link below is to the env file. It still does not work. When I connect to it, it shows connection refused. Can you see what I might not have configured properly?

https://drive.google.com/file/d/1e1lOMylj2zC8Dse6uaf33UIFDu2fuxzQ/view?usp=sharing

When you say DB connection credentials in the LTI setting, I am not sure which database. I thought it uses the default settings from the docker.

Thank you.
